fix false positives in constParameter

shaneasd opened the pull request:
Trying to upgrade my codebase from cppcheck 1.90 to 2.1 I encountered a handful of new constParameter warnings in methods that looked something like
```
void mymethod(T& x)
{
   U& y = dynamic_cast<U&>(x);
   dostuffwithy(y);
}
```
which would warn that x can be const. Making x const would cause the cast to fail because it would lose qualifiers.

I've added a unit test for this case and, I believe, fixed it. I also realised that a simple up-casting assignment of the form ```U& y = x;``` would also not be detected so fixed that case as well. My focus has been on parameters only. I suspect my fix applies to variables in general but I have not attempted to verify this.

pfultz2 commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            auto match = Token::findmatch(var->nameToken(), "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId());
```
An assignment to a reference should be checked in the `isVariableChanged` function instead.

shaneasd replied:
actually just to be clear, if x is the variable in question the expression we're detecting here is y = x, not x = y. Do you still think it should be considered modifying the variable? It does allow the variable to be modified but that's basically what all these checks are looking for.

pfultz2 replied:
Actually, `isVariableChanged` already checks if its assigned to a non-const reference [here](https://github.com/danmar/cppcheck/blob/main/lib/astutils.cpp#L1477). The extra check here should not be necessary unless there is some bug in the code.

shaneasd replied:
That's not the case I'm testing for here (I think). Suppose I start with ```X& x```.
If I say ```x = _``` then isVariableChanged should (and does as far as I can tell) detect that x has been assigned to and thus cannot be const.

The case I'm trying to detect is shown here https://godbolt.org/z/ze8r8o
```xc``` cannot be const because it prevents y2 from compiling. More generally a reference from which a non-const reference is created cannot itself be const. Interestingly, I think I'm failing to respect the fact that a const& is still fine here. I'll add a test for that and attempt a fix.

pfultz2 replied:
In that case `y2` can be `const`(and we should warn about this), and when that is changed to `const` then `xc` can be `const` as well. I dont really see that as an FP since we should warn about all the variables that can be declared `const`.

shaneasd replied:
that depends on how y2 is used. If we call a non-const method on y2 (as I do in the unit tests) then it can't be const. If we don't then it can and once we fix the ensuing warning we'll then get the warning about xc being const.

shaneasd replied:
>I think I'm failing to respect the fact that a const& is still fine here. I'll add a test for that and attempt a fix.

I was wrong about this. Looks like the test and code handle it.

pfultz2 replied:
> If we call a non-const method on y2 (as I do in the unit tests) then it can't be const.

But this should be caught by `isVariableChanged`, if not there is a bug in `isVariableChanged`.

shaneasd replied:
should isVariableChanged return true if we call y2.mutate() and can't see the definition of y2.mutate()? i.e. does this method indicate we know the variable has changed or just that we can't know that it hasn't? I assumed the former.

pfultz2 commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
+                    continue;
+        }
+        // Skip if we ever dynamic_cast/static_cast this variable to a non-const type
+        if (Token::findmatch(var->nameToken(), "dynamic_cast|static_cast < %type% & > ( %varid% )", scope->bodyEnd, var->declarationId()))
```
How would this handle `dynamic_cast<typename T::U&>(x)` or `dynamic_cast<T*>(&x)`?

shaneasd replied:
There's a case already for taking the address of the variable which I would expect to capture ```dynamic_cast<T*>(&x)``` but I can add a test for that regardless.

It probably won't capture ```dynamic_cast<typename T::U&>(x)``` currently. Is there a way to add optional tokens to the match string or should I just test against two different strings: one with and one without typename?

shaneasd replied:
I've added tests for both of these cases and modified the pattern to optionally include typename

pfultz2 replied:
I wonder if it would be better to traverse the ast, but that requires a little more work because we dont have any ast matchers.

shaneasd replied:
I think I prefer it this way but I could be convinced.

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
+        // Skip if another non-const variable is initialized with this variable
+        {
+            auto match = Token::findmatch(var->nameToken(), "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId());
+            if (match)
```
I suggest:

    const Variable *matchvar = match ? match->tokAt(2)->variable() : nullptr;
    if (matchvar && !matchvar->isConst())
        continue;

that would handle extended declarations better like `const static unsigned int & x = y;`

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            auto match = Token::findmatch(var->nameToken(), "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId());
```
maybe it would be better to loop through all tokens. imagine that there is first a const reference initialization and then a non-const initialization. This code will only check the first initialization.

shaneasd replied:
I've extended my tests to include this case and indeed it currently fails.

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
+                    continue;
+        }
+        // Skip if we ever dynamic_cast/static_cast this variable to a non-const type
+        if (Token::findmatch(var->nameToken(), "dynamic_cast|static_cast < typename| %type% & > ( %varid% )", scope->bodyEnd, var->declarationId()))
```
if a loop is used I think a better pattern might be:

    tok->isCast() && Token::Match(tok->tokAt(-2), "& > ( %varid% )", var->declarationId())

your code is pretty good but I think that a loop would be more robust.

shaneasd commented:
@danmar I've acted on your feedback

shaneasd commented:
Looks like my change might have tripped one of the tests for garbage code. I'll need to add some null checks I guess.

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
+        {
+            //Is it the right side of an initialization of a non-const reference
+            bool usedInAssignment = [&]
+                {
```
Imho, this outer lambda should be removed.

shaneasd replied:
Originally I didn't have it but I ended up wanting to ```continue``` the outer for loop from inside the new inner one which I can't do. However I liked the pattern of breaking out as soon as you know you've got something so I used return instead. That meant I needed something to return from.

I could instead have started with ```bool usedInAssignment = false;``` and then set it to true in my loop and added an extra condition on usedInAssignment to the loop in order to break out early. Personally I would find that code harder to read as it's less clear that the whole point of that block of code is to calculate this one boolean and it would mean the boolean would exist in an intermediate state which I also don't like.

Let me know your thoughts on this and if you insist then I'll change it.

shaneasd replied:
actually since I can just copy-paste your solution from below I've done that

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
+            bool usedInAssignment = [&]
+                {
+                    auto findAssignment = [&](const Token* t) { return Token::findmatch(t, "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId()); };
+                    for (const Token* match = findAssignment(var->nameToken()); match != nullptr; match = findAssignment(match->next()))
```
I would personally prefer:

    bool usedInAssignment = false;
    for (const Token *tok = var->nameToken(); tok != scope->bodyEnd; tok = tok->next()) {
        if (!Token::Match(tok, "& %var% = %varid%", var->declarationId())
            continue;
        const Variable *refvar = tok->next()->variable();
        if (refvar && !refvar->isConst() && refvar->nameToken() == tok->next()) {
            usedInAssignment = true;
            break;
        }
    }

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
+                            {
+                                // dynamic_cast<X&>(y)
+                                //                 ^    we're at this operator
+                                bool castToConst = Token::simpleMatch(tok->previous()->link()->next(), "const");
```
can you please try to use the ValueType:

    if (tok->valueType() && (tok->valueType()->constness & 1))

I am not sure if we want to test `constness & 1` or `constness > 1`. What should happen for `dynamic_cast<X * const>`?

shaneasd replied:
I don't think ```dynamic_cast<X * const>``` makes any sense. It's already an rvalue. What difference does it make if it's const? You can freely say ```B* b = dynamic_cast<B * const>(x)``` but can't say ```B* b = dynamic_cast<B const *>(x)```

It's not clear to me what valueType represents or how the bits of constness should be interpreted. Can you explain?

danmar commented on lib/checkother.cpp:
```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
+                            {
+                                // (X&)(y)
+                                // ^        we're at this operator
+                                bool castToConst = Token::simpleMatch(tok->next(), "const");
```
I believe, if `valueType` is used we shouldn't need separate handling for C and C++ casts.

danmar reviewed (commented):
Thanks for working on this.. some more nits..

shaneasd commented:
I've made my initial attempt at including the concept of a reference in valueType. It's not easy and I'm not confident I've done it well.

I'm reasonably happy with the design of having an extra bool indicating whether the top level indirection is a reference or a pointer and the actual generation of the valuetype including references is not that hard. What concerns me is the hundreds of places that use the ```pointer``` member to mean something which it does not technically mean anymore. My strategy has been to rename the pointer member and add methods to query whether the type is a pointer, a reference or either. Then I went through and corrected all references to ```pointer``` with either the renamed member or one of the three new methods as appropriate. For some cases it was quite obvious which of the 4 options to use. For many it was not. The end result is 29 test failures at the moment.

I've pushed this change for review to give an idea of the scope of the problem and get suggestions on how to address it in a more controlled manner. I don't like creating a pull request with 100s of corrections and I think it's pretty impossible to review.

danmar commented on lib/checkother.cpp:
```diff
@@ -1408,34 +1408,45 @@ void CheckOther::checkConstVariable()
                     {
                         if (tok->isCast())
                         {
+                            const Token* typeStart = nullptr; //first token of the type we're casting to
```
I think this `typeStart` and `typeEnd` is an interesting approach .. it could be an acceptable first step until we have improved `valueType`.

shaneasd replied:
I actually wrote this before I started trying to fix typeValue. It's just pushed here so I don't lose it but it will hopefully be unnecessary if I use typeValue. It's actually not as effective as I thought it was since figuring out the complexities of the type token sequence is far from trivial even in this fairly simple case.

danmar commented on lib/symboldatabase.h:
```diff
@@ -1151,7 +1151,8 @@ class CPPCHECKLIB ValueType {
     nonneg int bits;                    ///< bitfield bitcount
-    nonneg int pointer;                 ///< 0=>not pointer, 1=>*, 2=>**, 3=>***, etc
+    nonneg int pointerDepth;            ///< 0=>not pointer, 1=>*, 2=>**, 3=>***, etc
+    bool reference;                     ///< Is the outermost pointer actually a reference. i.e. if pointer==2 then are we dealing with an X** or an X*&
```
hmm.. how about `X&`  => pointerDepth=0, reference=true
I am not sure how that will affect all the code in practice I believe you have a better idea. It just seems a bit more intuitive? To calculate the type for a `*p` expression we always decrease `pointerDepth` with 1 and clear `reference`.

shaneasd replied:
I think that's a very good suggestion. I think it will allow me to limit the scope of my changes so that we can add the knowledge of references incrementally rather than being forced to update everything. Of course that probably means we'll miss a bunch of existing bugs where references aren't handled properly but it's a lot less work and much less likely to introduce new bugs.

danmar commented on lib/symboldatabase.h:
```diff
@@ -1160,11 +1161,24 @@ class CPPCHECKLIB ValueType {
     std::string originalTypeName;         ///< original type name as written in the source code. eg. this might be "uint8_t" when type is CHAR.
 
+    bool isPointer() const {
+        return pointerDepth > 0 && !reference;
```
in my humble opinion a `X*&` is a pointer.

danmar replied:
I guess it's just a question about naming? The name `isNonReferencePointer()` would be more clear but it is rather long..

danmar reviewed (commented):
I think your updates to `valueType` are interesting. but yes as you say it's probably better to merge those separately. Your WIP fix with a `typeStart` and `typeEnd` looks good to me.

shaneasd commented on test/testother.cpp:
```diff
@@ -2051,6 +2051,187 @@ class TestOther : public TestFixture {
+            "    const U const * const * const * const y = dynamic_cast<const U const * const * const * const>(&x)\n"
+            "    y->mutate();\n" //to avoid warnings that y can be const
+            "}");
+        TODO_ASSERT_EQUALS("can be const", errout.str(), ""); //Currently taking the address is treated as a non-const operation when it should depend on what we do with it
```
My code no longer cares whether you're casting to a pointer or reference but it won't actually be executed in this case because taking the address is enough to treat it as non-const.

shaneasd replied:
And if the original variable is a pointer then it will fail the isReferenceCheck at https://github.com/danmar/cppcheck/blob/8e79b0c8bc825d94446432f4d00c761e395b85c7/lib/checkother.cpp#L1341

shaneasd commented:
I'm reasonably happy with this solution now. It produces one test failure:
```
P:\cppcheck\test\testcondition.cpp:2340(TestCondition::oppositeInnerConditionEmpty): Assertion failed.
Expected:
[test.cpp:1] -> [test.cpp:1]: (warning) Opposite inner 'if' condition leads to a dead code block.\n

Actual:

```

The test code is ```template<class T> void f1(const T &s) { if(s.size() > 42) if(s.empty()) {}}```

The difference is the ```|| valuetype->reference``` added in symbol database. s now has a valueType which it did not before. Ultimately this means we don't use the fallback in https://github.com/danmar/cppcheck/blob/6c588cc3efe4bf14f7c4d4dd503dceef8b26f16a/lib/library.cpp#L1149

I think it's arguably correct for the code to not warn here. It doesn't know what the type of T is so it can't know the meaning of size and empty

danmar commented:
> I think it's arguably correct for the code to not warn here. It doesn't know what the type of T is so it can't know the meaning of size and empty

It sounds reasonable to have this false negative. Cppcheck can't warn unless the template is instantiated.

pfultz2 commented on lib/symboldatabase.h:
```diff
@@ -1153,6 +1153,7 @@ class CPPCHECKLIB ValueType {
     nonneg int bits;                    ///< bitfield bitcount
     nonneg int pointer;                 ///< 0=>not pointer, 1=>*, 2=>**, 3=>***, etc
     nonneg int constness;               ///< bit 0=data, bit 1=*, bit 2=**
+    bool reference = false;             ///< Is the outermost indirection of this type a reference
```
We should distinguish between a lvalue and rvalue reference. This should probably be an enum.

shaneasd replied:
I agree. I've made this change and updated the parsing code to detect rvalue references. At this stage nothing uses RValue references but the option is there.

danmar commented:
I feel more work is needed for the references:

1. Please tweak `ValueType::str` in lib/symboldatabase.cpp to ensure that references are shown. I have the feeling you can test that with `--debug` look at the type info in the ASTs.

2. Add some tests in `TestSymbolDatabase::valuetype`. Variable declarations, calculations, ... I feel that this should work already with existing code: `refvar1+refvar2 => no reference` but some such test should be added. are there some calculations where ref should be preserved? I guess the reference should be preserved in assignment result / cast result.

3. Take a look at `--dump` output. The `ValueType::dump` should be updated in lib/symboldatabase.cpp.

shaneasd commented:
1. updated str and verified with --debug -v
2. will do
3. updated --dump and verified in the --dump file.

danmar commented on lib/symboldatabase.cpp:
```diff
@@ -6352,6 +6352,13 @@ std::string ValueType::dump() const
         ret << " valueType-constness=\"" << constness << '\"';
 
+    if (reference == Reference::None)
+        ret << " valueType-reference=None";
```
this needs to be a valid xml attribute: `valueType-reference="None"`.

shaneasd commented:
I've fixed the xml and added tests for variable declarations. I'm not sure about preserving the reference. I guess canonically assignment will return a reference. Dereferencing will as well of course. Other than that nothing springs to mind. What do you imagine the test looking like?

danmar commented:
In TestSymbolDatabase::valueType there is this:

        // Variable calculations
        ASSERT_EQUALS("void *", typeOf("void *p; a = p + 1;", "+"));

somewhere below that there should be a few tests where references are used. Imho it's enough with 1 test where a reference variable is used in assignment lhs. Maybe this will work:

        ASSERT_EQUALS("int &", typeOf("int &ref; ref = 123;", "="));
        ASSERT_EQUALS("int", typeOf("int &ref; x = ref << 1", "<<"));
        ASSERT_EQUALS("int", typeOf("int &ref; x = ref + ref;", "+"));

I have the feeling that references would mostly be handled properly in calculations because the default value is none.

shaneasd commented:
Sounds good. I've added something like that. Unsurprisingly the dereference operator yields a value not a reference so I'll have to fix that tomorrow.

danmar commented on test/testsymboldatabase.cpp:
```diff
@@ -6804,6 +6804,9 @@ class TestSymbolDatabase: public TestFixture {
         ASSERT_EQUALS("A::BC *", typeOf("namespace A { struct BC { int b; int c; }; }; struct A::BC *abc; x=abc+1;", "+"));
+        ASSERT_EQUALS("signed int", typeOf("auto a(int& x, int& y) { return x + y; }", "+"));
+        ASSERT_EQUALS("signed int &", typeOf("void a(int& x, int& y) { x = y; }", "="));
+        ASSERT_EQUALS("signed int &", typeOf("auto a(int* y) { return *y; }", "*"));
```
unfortunately I believe this will test the first `"*"`. I think it's a bit unfortunate if the valuetype is set in a variable declaration but I guess.

I am not sure but shouldn't the type be "signed int" when you dereference a "signed int *"?

shaneasd replied:
What is the best choice for cppcheck I can't say but in c++ the answer is no. For demonstration see: https://godbolt.org/z/Ms1qz6

For a logical justification, consider the statement ```*x = 4;``` if *x produces a T rather than a T& then the assignment does nothing.

It occurs to me that the rule is probably different in c where references do not exist. Does any action need to be taken to address this? I haven't considered the c use case at all in this PR as I never write c. I guess most of the changes here should be conditional on the language? At the very least the answer to this specific question seems to be different.

danmar replied:
hmm.. it seems logical to be consistent with the clast ast dump. And in that neither `*p` or `=` results are references.

Your test code:
```
danielm@edge:~/cppcheck$ cat 1.cpp 
auto foo(int *p)
{
    return *p;
}
danielm@edge:~/cppcheck$ clang -cc1 -ast-dump 1.cpp
TranslationUnitDecl 0x974428 <<invalid sloc>> <invalid sloc>
.....
`-FunctionDecl 0x9b3ef8 <1.cpp:1:1, line:4:1> line:1:6 foo 'int (int *)'
  |-ParmVarDecl 0x9b3dd8 <col:10, col:15> col:15 used p 'int *'
  `-CompoundStmt 0x9b4200 <line:2:1, line:4:1>
    `-ReturnStmt 0x9b41f0 <line:3:5, col:13>
      `-ImplicitCastExpr 0x9b41d8 <col:12, col:13> 'int' <LValueToRValue>
        `-UnaryOperator 0x9b4058 <col:12, col:13> 'int' lvalue prefix '*' cannot overflow
          `-ImplicitCastExpr 0x9b4040 <col:13> 'int *' <LValueToRValue>
            `-DeclRefExpr 0x9b4020 <col:13> 'int *' lvalue ParmVar 0x9b3dd8 'p' 'int *'
```

A reference variable "ref" has the type `int &`. But when assigning to the ref variable the `=` has type `int`..
```
danielm@edge:~/cppcheck$ cat 1.cpp 
void foo(int *p)
{
    int &ref = *p;
    ref = 123;
}
danielm@edge:~/cppcheck$ clang -cc1 -ast-dump 1.cpp
TranslationUnitDecl 0x1989428 <<invalid sloc>> <invalid sloc>
.....
`-FunctionDecl 0x19c8ec0 <1.cpp:2:1, line:6:1> line:2:6 foo 'void (int *)'
  |-ParmVarDecl 0x19c8df8 <col:10, col:15> col:15 used p 'int *'
  `-CompoundStmt 0x19c9180 <line:3:1, line:6:1>
    |-DeclStmt 0x19c9108 <line:4:5, col:18>
    | `-VarDecl 0x19c8ff8 <col:5, col:17> col:10 used ref 'int &' cinit
    |   `-UnaryOperator 0x19c9098 <col:16, col:17> 'int' lvalue prefix '*' cannot overflow
    |     `-ImplicitCastExpr 0x19c9080 <col:17> 'int *' <LValueToRValue>
    |       `-DeclRefExpr 0x19c9060 <col:17> 'int *' lvalue ParmVar 0x19c8df8 'p' 'int *'
    `-BinaryOperator 0x19c9160 <line:5:5, col:11> 'int' lvalue '='
      |-DeclRefExpr 0x19c9120 <col:5> 'int' lvalue Var 0x19c8ff8 'ref' 'int &'
      `-IntegerLiteral 0x19c9140 <col:11> 'int' 123
```

.. I am a bit surprised. But well right now I have the feeling it would be a good idea to be consistent.

If a reference is used or not that will not be a direct problem for `ValueFlow` etc because we really need to handle C code also so we must have this logic when assigning to non-references etc.

shaneasd replied:
ok so it sounds like we want the result of all three of these operations to be a value rather than a reference. I think that's technically wrong and may cause bugs but they'll be bugs that already exist and fixing them may cause new bugs so I'm happy to leave it.

Is that the only concern? Are there more tests you want to see?

danmar replied:
> Are there more tests you want to see?

I think this is it.

danmar replied:
clang ast clearly says that some values are "lvalue" . this info is not stored in cppcheck ast (yet).

pfultz2 replied:
> hmm.. it seems logical to be consistent with the clast ast dump. And in that neither *p or = results are references.

The clang ast shows its an lvalue:

> `-DeclRefExpr 0x9b4020 <col:13> 'int *' lvalue ParmVar 0x9b3dd8 'p' 'int *'

For C++, that would be considered a reference. Instead of describing the value category in terms of C++ references, which dont exist in C, we should use the actual value categories: lvalue(`Reference::LValue`), prvalue(`Reference::None`), and xvalue(`Reference::RValue`). With the exception of xvalue(which is not possible in C), these should be the same for both C and C++ programs.

`ValueType::str()` could work more like `typeof` which strips the reference rather than like `decltype`. That should less confusing for C, which doesn't have references.

danmar replied:
thanks for looking... if it makes sense to add the lvalue/rvalue info I am not against that. but I do not feel that it should replace the reference info that we have now. Clang has both such reference info and also if the value is a lvalue.

The type of a variable is not lvalue/rvalue. The result from `Variable::valueType()` should not say lvalue/rvalue.

shaneasd replied:
I pushed and then reverted d73bafa which was my work in progress attempt to have operations return l/r-value references "correctly". If you decide to pursue this it might be a useful starting point. I believe it was mostly working.

danmar commented:
ok thanks I merged it now. I am eager to test it out! have you been added to our AUTHORS file?

shaneasd commented:
yep I'm already there. Thanks
